### PR TITLE
PSP-9513 : Change document workflow - support asynchronous upload

### PR DESCRIPTION
**Pull request by eddherrera**

### CODE REVIEW

File: source/backend/api/Services/FormDocumentService.cs [hunk 1]
```diff
@@ -60,6 +60,47 @@ public IList<PimsFormType> GetFormDocumentTypes(string formTypeCode)
             return new List<PimsFormType>() { _formTypeRepository.GetByFormTypeCode(formTypeCode) };
         }
 
+        public async Task UploadFormDocumentTemplate(string formTypeCode, DocumentUploadRequest uploadRequest)
+        {
+            throw new NotImplementedException();
```
Comment (devinleighsmith) [inline source/backend/api/Services/FormDocumentService.cs:65]: Can you explain what is going on here? generally we do no leave blocks of commented-out code in the application.

File: source/backend/api/Services/DocumentFileService.cs [hunk 2]
```diff
@@ -89,333 +91,509 @@ public IList<T> GetFileDocuments<T>(FileType fileType, long fileId)
...
-                    ResearchFileId = researchFileId,
-                    DocumentId = uploadResult.Document.Id,
-                };
-                newResearchFileDocument = researchFileDocumentRepository.AddResearch(newResearchFileDocument);
-                researchFileDocumentRepository.CommitTransaction();
-
-                relationshipResponse.DocumentRelationship = mapper.Map<DocumentRelationshipModel>(newResearchFileDocument);
-            }
+                DocumentId = pimsDocument.DocumentId,
+                AcquisitionFileDocumentId = newAcquisitionDocument.AcquisitionFileDocumentId,
```
Comment (devinleighsmith) [inline source/backend/api/Services/DocumentFileService.cs:119]: You can take this line out, as we are removing all file foreign key relationships from the document queue table. Same goes for each of the file types.

File: source/backend/api/Services/DocumentFileService.cs [hunk 3]
```diff
@@ -89,333 +91,509 @@ public IList<T> GetFileDocuments<T>(FileType fileType, long fileId)
...
-            DocumentUploadRelationshipResponse relationshipResponse = new DocumentUploadRelationshipResponse()
+            // Step 2 - Create the File Document
+            PimsAcquisitionFileDocument newAcquisitionDocument = new()
             {
-                UploadResponse = uploadResult,
+                AcquisitionFileId = acquisitionFileId,
+                DocumentId = pimsDocument.DocumentId,
             };
+            _acquisitionFileDocumentRepository.AddAcquisition(newAcquisitionDocument);
+            _documentQueueRepository.SaveChanges();
```
Comment (devinleighsmith) [inline source/backend/api/Services/DocumentFileService.cs:113]: I don't think you want to call save changes between repository transactions, I'm guessing all of this work should be done in a single transaction to allow rollback if a later transaction fails.

File: source/backend/api/Helpers/Extensions/DocumentUploadRequestExtension.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+using System;
+using Pims.Api.Models.Requests.Document.Upload;
+using Pims.Core.Api.Exceptions;
+
+namespace Pims.Api.Helpers.Extensions
+{
+    public static class DocumentUploadRequestExtension
+    {
+        public static void ThrowInvalidFileSize(this DocumentUploadRequest documentUploadRequest)
```
Comment (devinleighsmith) [inline source/backend/api/Helpers/Extensions/DocumentUploadRequestExtension.cs:9]: Hmm, I'm curious about what the benefit of moving this to an extension method is, can you provide some background?

File: source/backend/api/Services/DocumentFileService.cs [hunk 5]
```diff
@@ -89,333 +91,509 @@ public IList<T> GetFileDocuments<T>(FileType fileType, long fileId)
...
-                UploadResponse = uploadResult,
+                ProjectId = projectId,
+                DocumentId = pimsDocument.DocumentId,
             };
+            _projectRepository.AddProjectDocument(newFileDocument);
+            _documentQueueRepository.SaveChanges();
 
-            // Throw an error if Mayan returns a null document. This means it wasn't able to store it.
-            ValidateDocumentUploadResponse(uploadResult);
+            throw new NotImplementedException();
```
Comment (devinleighsmith) [inline source/backend/api/Services/DocumentFileService.cs:193]: @eddherrera what is going on here? why can't we save project documents?

File: source/backend/api/Services/DocumentFileService.cs [hunk 6]
```diff
@@ -89,333 +91,509 @@ public IList<T> GetFileDocuments<T>(FileType fileType, long fileId)
...
             }
+
+            return result;
         }
 
-        private static void ValidateZeroLengthFile(DocumentUploadRequest uploadRequest)
+        private PimsDocument RemoveDocumentMayanID(PimsDocument doc)
         {
-            if (uploadRequest.File is not null && uploadRequest.File.Length == 0)
+            // Update the Mayan ID from PIMS, this will leave a trace
```
Comment (devinleighsmith) [inline source/backend/api/Services/DocumentFileService.cs:568]: I don't think this method executed alone will leave a trace.

File: source/backend/api/Services/DocumentFileService.cs [hunk 7]
```diff
@@ -89,333 +91,509 @@ public IList<T> GetFileDocuments<T>(FileType fileType, long fileId)
...
             Logger.LogInformation("Deleting PIMS document for single research file");
             User.ThrowIfNotAllAuthorized(Permissions.DocumentDelete, Permissions.ResearchFileEdit);
 
-            var relationshipCount = _documentRepository.DocumentRelationshipCount(researchFileDocument.DocumentId);
-            if (relationshipCount == 1)
-            {
-                return await documentService.DeleteDocumentAsync(researchFileDocument.Document);
-            }
-            else
+            var result = new ExternalResponse<string>() { Status = ExternalResponseStatus.NotExecuted };
```
Comment (devinleighsmith) [inline source/backend/api/Services/DocumentFileService.cs:339]: hmm, if a status of NotExecuted is returned in the case of a document with no mayanId, does that cause problems on the frontend?

File: source/backend/api/Services/DocumentFileService.cs [hunk 8]
```diff
@@ -89,333 +91,509 @@ public IList<T> GetFileDocuments<T>(FileType fileType, long fileId)
...
+            return _documentRepository.Update(doc, false);
+        }
+
+        private void DeleteQueuedDocumentItem(long documentId)
+        {
+            var documentQueuedItem = _documentQueueRepository.GetByDocumentId(documentId);
+            if (documentQueuedItem.DocumentQueueStatusTypeCode == DocumentQueueStatusTypes.PENDING.ToString()
+                || documentQueuedItem.DocumentQueueStatusTypeCode == DocumentQueueStatusTypes.PROCESSING.ToString())
+            {
+                throw new BadRequestException("Doucment in process can not be deleted");
```
Comment (devinleighsmith) [inline source/backend/api/Services/DocumentFileService.cs:579]: *Document

File: source/backend/apimodels/Models/Concepts/Document/DocumentMap.cs [hunk 9]
```diff
@@ -14,6 +15,7 @@ public void Register(TypeAdapterConfig config)
                 .Map(dest => dest.DocumentType, src => src.DocumentType)
                 .Map(dest => dest.StatusTypeCode, src => src.DocumentStatusTypeCodeNavigation)
                 .Map(dest => dest.FileName, src => src.FileName)
+                .Map(dest => dest.DocumentQueueStatusTypeCode, src => src.PimsDocumentQueues.Count > 0 ? src.PimsDocumentQueues.FirstOrDefault().DocumentQueueStatusTypeCodeNavigation : null)
```
Comment (devinleighsmith) [inline source/backend/apimodels/Models/Concepts/Document/DocumentMap.cs:18]: This is an interesting one, as generally we try to pass concepts as-is without merging entity fields. Thoughts @FuriousLlama ?

File: source/backend/dal/Repositories/DocumentRepository.cs [hunk 10]
```diff
@@ -35,7 +35,7 @@ public DocumentRepository(
         #region Methods
 
         /// <summary>
-        /// Get the document from the database based on document id.
+        /// Get the document from the database based on document id.z
```
Comment (devinleighsmith) [inline source/backend/dal/Repositories/DocumentRepository.cs:38]: z?

**Comment by eddherrera** [comment]
@devinleighsmith I am changing this one to a draft since it seems it would still need adjustments with the pending changes coming to IS96 DB Schema. I'll address most of the comments now.

**Comment by devinleighsmith** [comment]
@eddherrera I've done some local testing and I don't see the processing or error states (of the document queue) being shown on the frontend

**Comment by devinleighsmith** [comment]
@eddherrera I don't think any of my comments need to wait for db changes, they relate to changing the logic to better fit both the current and the upcoming schema. There should be no upcoming schema that you need to wait for (in regards to my changes at least).

File: source/frontend/src/components/common/buttons/RefreshButton.tsx [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * Button displaying a refresh/recycle icon, used to reset form data.
+ * @param param0
+ */
+const RefreshButton: React.FC<React.PropsWithChildren<ButtonProps>> = ({ ...props }) => {
+  return (
+    <TooltipWrapper tooltipId="btn-refresh-tooltip" tooltip="Refresh">
+      <Button
+        data-testid="refresh-button"
```
Comment (devinleighsmith) [inline source/frontend/src/components/common/buttons/RefreshButton.tsx:15]: given that this button will be placed inside other screens, please make this id/testid/title more specific

File: source/frontend/src/features/mapSideBar/acquisition/models/PayeeOptionModel.ts [hunk 12]
```diff
@@ -102,7 +102,6 @@ export class PayeeOption {
       agreementDate: null,
       expropriationNoticeServedDate: null,
       expropriationVestingDate: null,
-      advancedPaymentServedDate: null,
```
Comment (devinleighsmith) [inline source/frontend/src/features/mapSideBar/acquisition/models/PayeeOptionModel.ts:105]: hmm, there seems to be some unrelated changes in this PR, thoughts?

**Comment by devinleighsmith** [comment]
@eddherrera it seems that right now, when not metadata is included with the document, the word "null" is being written into the DOCUMENT_METADATA column.

File: source/backend/api/Services/FormDocumentService.cs (hunk 1 above)
Comment (eddherrera) [reply source/backend/api/Services/FormDocumentService.cs:65]: removed, this was part of draft.

File: source/backend/api/Services/DocumentFileService.cs (hunk 2 above)
Comment (eddherrera) [reply source/backend/api/Services/DocumentFileService.cs:119]: removed

File: source/backend/api/Services/DocumentFileService.cs (hunk 3 above)
Comment (eddherrera) [reply source/backend/api/Services/DocumentFileService.cs:113]: this is still a single transaction, SaveChanges() won't commit any work and it will rollback if the transaction fails.

File: source/backend/api/Helpers/Extensions/DocumentUploadRequestExtension.cs (hunk 4 above)
Comment (eddherrera) [reply source/backend/api/Helpers/Extensions/DocumentUploadRequestExtension.cs:9]: this private method was first defined on a different Service and since I needed it on another it made sense to put it as an Extension method at that time. At this point if it's only used in one service I could make it back to a private method.

File: source/backend/api/Services/DocumentFileService.cs (hunk 5 above)
Comment (eddherrera) [reply source/backend/api/Services/DocumentFileService.cs:193]: This was related to a column missing in the Queue Table but that has been fixed now. All good now.

File: source/backend/api/Services/DocumentFileService.cs (hunk 6 above)
Comment (eddherrera) [reply source/backend/api/Services/DocumentFileService.cs:568]: making sure that the mayan Id is deleted is the calling to mayan api is successful and making sure that mayan id is null after that so it does not point to a deleted file. trace is just the history table

File: source/backend/api/Services/DocumentFileService.cs (hunk 7 above)
Comment (eddherrera) [reply source/backend/api/Services/DocumentFileService.cs:339]: no issues.

File: source/backend/dal/Repositories/DocumentRepository.cs (hunk 10 above)
Comment (eddherrera) [reply source/backend/dal/Repositories/DocumentRepository.cs:38]: removed

File: source/frontend/src/components/common/buttons/RefreshButton.tsx (hunk 11 above)
Comment (eddherrera) [reply source/frontend/src/components/common/buttons/RefreshButton.tsx:15]: updated

File: source/frontend/src/features/mapSideBar/acquisition/models/PayeeOptionModel.ts (hunk 12 above)
Comment (eddherrera) [reply source/frontend/src/features/mapSideBar/acquisition/models/PayeeOptionModel.ts:105]: needed to compile, not sure who's fixing those?

File: source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx [hunk 13]
```diff
@@ -112,46 +132,103 @@ const renderActions = (
...
-              icon={<FaEye size={24} id={`document-view-${index}`} title="document view details" />}
-              onClick={() => original?.id && onViewDetails(DocumentRow.toApi(original))}
-            ></Button>
-          </Col>
-        )}
-        {hasClaim(Claims.DOCUMENT_DELETE) && (
-          <Col>
+
+    const documentInError =
+      (original.mayanDocumentId === null || original.mayanDocumentId === -1) &&
```
Comment (devinleighsmith) [inline source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx:137]: This logic appears inconsistent across this file. Also, I believe you can use isValidId to be consistent with the rest of our code base.

Comment (devinleighsmith) [reply source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx:137]: In my testing, mayanId is being converted from null to 0 on the backend, so none of these icons are being displayed.

Comment (devinleighsmith) [reply source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx:137]: Also, it looks like some of these boolean statements can be collapsed to make them more readable (ie: original.queueStatusTypeCode && original.queueStatusTypeCode.id === ApiGen_CodeTypes_DocumentQueueStatusTypes.PENDING
is equivalent to original?.queueStatusTypeCode === ApiGen_CodeTypes_DocumentQueueStatusTypes.PENDING

File: source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx [hunk 14]
```diff
@@ -112,46 +132,103 @@ const renderActions = (
...
+            icon={<FaEye size={21} id={`document-view-${index}`} title="document view details" />}
+            onClick={() => original?.id && onViewDetails(DocumentRow.toApi(original))}
+          ></Button>
+        )}
+
+        {hasClaim(Claims.DOCUMENT_DELETE) && (
+          <StyledRemoveLinkButton
+            data-testid="document-delete-button"
+            icon={<FaTrash size={21} id={`document-delete-${index}`} title="document delete" />}
+            onClick={() => original?.id && onDelete(DocumentRow.toApi(original))}
```
Comment (devinleighsmith) [inline source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx:204]: I'd prefer to not display the button if there is no id, instead of displaying a button that does nothing when clicked.

File: source/backend/api/Services/DocumentFileService.cs (hunk 3 above)
Comment (devinleighsmith) [reply source/backend/api/Services/DocumentFileService.cs:111]: at the very least you don't need to call savechanges twice, have a look at that method.

File: source/backend/dal/Repositories/DocumentQueueRepository.cs [hunk 15]
```diff
@@ -32,6 +34,36 @@ public DocumentQueueRepository(
...
+        }
+
+        /// <summary>
+        /// Find Queue Item for a Document.
+        /// </summary>
+        /// <param name="documentId"></param>
+        /// <returns></returns>
+        public PimsDocumentQueue GetByDocumentId(long documentId)
+        {
+            return Context.PimsDocumentQueues.Where(x => x.DocumentId == documentId).FirstOrDefault();
```
Comment (devinleighsmith) [inline source/backend/dal/Repositories/DocumentQueueRepository.cs:64]: should this be asNoTracking?

File: source/backend/tests/unit/api/Services/DocumentFileServiceTest.cs [hunk 16]
```diff
@@ -288,48 +290,48 @@ public void UploadDocumentAsync_Disposition_ShouldThrowException_NotAuthorized()
...
-                DocumentStatusCode = "DocumentStatus",
-            };
-
-            await service.UploadProjectDocumentAsync(1, uploadRequest);
-
-            // Assert
-            documentService.Verify(x => x.UploadDocumentAsync(It.IsAny<DocumentUploadRequest>()), Times.Once);
-            projectRepository.Verify(x => x.AddProjectDocument(It.IsAny<PimsProjectDocument>()), Times.Once);
-        }
+        //[Fact]
```
Comment (devinleighsmith) [inline source/backend/tests/unit/api/Services/DocumentFileServiceTest.cs:300]: @eddherrera why is this commented out?

File: source/frontend/src/components/common/buttons/RefreshButton.tsx [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+ */
+const RefreshButton: React.FC<React.PropsWithChildren<IRefreshButtonProps>> = ({
+  toolId,
+  toolText,
+  refreshButtonId,
+  refreshButtonTestId,
+  refreshButtonTitle,
+  ...props
+}) => {
+  const refreshButtonIdValue = refreshButtonId ?? 'refresh-button';
```
Comment (devinleighsmith) [inline source/frontend/src/components/common/buttons/RefreshButton.tsx:32]: please set the defaults such that the dom won't throw an error due to duplicate ids.

File: source/frontend/src/features/documents/documentUpload/DocumentUploadContainer.tsx [hunk 18]
```diff
@@ -125,7 +125,7 @@ export const DocumentUploadContainer = forwardRef<
       await Promise.all(uploadDocumentTasks);
     setIsUploading(false);
     if (tasksResult !== undefined) {
-      const batchResults = tasksResult.map<BatchUploadResponseModel>((r, index) => {
+      const batchResults = tasksResult.map((r, index) => {
```
Comment (devinleighsmith) [inline source/frontend/src/features/documents/documentUpload/DocumentUploadContainer.tsx:128]: Why remove the type? typings are nice!

File: source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx [hunk 19]
```diff
@@ -112,46 +132,103 @@ const renderActions = (
...
+        original.queueStatusTypeCode.id === ApiGen_CodeTypes_DocumentQueueStatusTypes.MAYAN_ERROR);
+
+    const documentProcessing =
+      (original.mayanDocumentId === null || original.mayanDocumentId < 1) &&
+      original.queueStatusTypeCode &&
+      (original.queueStatusTypeCode.id === ApiGen_CodeTypes_DocumentQueueStatusTypes.PENDING ||
+        original.queueStatusTypeCode.id === ApiGen_CodeTypes_DocumentQueueStatusTypes.PROCESSING);
+
+    const canViewDocument =
+      (original.mayanDocumentId &&
```
Comment (devinleighsmith) [inline source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx:149]: the mayan document id being present isn't sufficient to download a document if it is still being processed - need to check for success.

File: source/backend/api/Services/DocumentFileService.cs (hunk 3 above)
Comment (eddherrera) [reply source/backend/api/Services/DocumentFileService.cs:111]: not sure what you mean, savechanges is called just 3 times, one for the document, for the file-document and for the que queue document.

File: source/backend/dal/Repositories/DocumentQueueRepository.cs (hunk 15 above)
Comment (eddherrera) [reply source/backend/dal/Repositories/DocumentQueueRepository.cs:64]: Needed to update the mayanId to NULL after successful delete call to Mayan API.

File: source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx (hunk 14 above)
Comment (eddherrera) [reply source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx:204]: A little lost here, the only way this displays is when the document has been processed as either success or failure. if it fails it deletes everything so they can retry.

File: source/frontend/src/features/documents/documentUpload/DocumentUploadContainer.tsx (hunk 18 above)
Comment (eddherrera) [reply source/frontend/src/features/documents/documentUpload/DocumentUploadContainer.tsx:128]: I'll try to put it back, not quite the type I am expecting since this one is tightly coupled to Mayan responses.

File: source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx (hunk 19 above)
Comment (eddherrera) [reply source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx:149]: I wouldn't expect a MayanId if it's still being processed, is this not the case?

File: source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx (hunk 14 above)
Comment (devinleighsmith) [reply source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx:198]: right now when the button is checked you ensure that original?.id is set before you fire the call to onDelete. I'd rather you either remove that check, or hide the entire button if original?.id is not set.

File: source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx (hunk 19 above)
Comment (devinleighsmith) [reply source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx]: Mayan ids are returned from mayan immediately so that you can track the progress of the document as it is processed. documents may have mayan ids but no file contents while they are being processed.

File: source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx (hunk 13 above)
Comment (eddherrera) [reply source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx]: Updated

File: source/backend/tests/unit/api/Services/DocumentFileServiceTest.cs (hunk 16 above)
Comment (eddherrera) [reply source/backend/tests/unit/api/Services/DocumentFileServiceTest.cs]: Updated

File: source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx (hunk 19 above)
Comment (eddherrera) [reply source/frontend/src/features/documents/list/DocumentResults/DocumentResultsColumns.tsx]: UIpdated

File: source/frontend/src/features/documents/documentUpload/DocumentUploadContainer.tsx (hunk 18 above)
Comment (eddherrera) [reply source/frontend/src/features/documents/documentUpload/DocumentUploadContainer.tsx]: updated

**Comment by eddherrera** [comment]
@devinleighsmith  please review comments and bugs are addressed

**Review by devinleighsmith — APPROVED** [review]
